src/java/org/apache/cassandra/index/sai/QueryContext.java

```diff
@@ -302,19 +309,88 @@ public Bits bitsetForShadowedPrimaryKeys(SegmentMetadata metadata, PrimaryKeyMap
                     break;
 
                 int ordinal = ordinalsView.getOrdinalForRowId(segmentRowId);
-                if (ordinal >= 0)
+                // row id does not have an ordinal in this segment.
+                if (ordinal < 0)
+                    continue;
+
+                if (graph.isOrdinalsToRowIdsIdentityMapping())
                 {
                     if (ignoredOrdinals == null)
                         ignoredOrdinals = new HashSet<>();
                     ignoredOrdinals.add(ordinal);
                 }
+                else
+                {
+                    if (ordinalsToRowIds == null)
+                        ordinalsToRowIds = new HashMap<>();
+                    ordinalsToRowIds.compute(ordinal, (k, rowIds) -> {
+                        if (rowIds == null)
+                            rowIds = new IntArrayList(1);
+                        rowIds.add(segmentRowId);
+                        return rowIds;
+                    });
+                }
             }
-        }
 
-        if (ignoredOrdinals == null)
-            return Bits.ALL;
+            // If we know we have an identity mapping, we do not have any partially shadowed ordinals.
+            if (graph.isOrdinalsToRowIdsIdentityMapping())
+                if (ignoredOrdinals != null)
+                    return Pair.create(new IgnoringBits(ignoredOrdinals, graph.size()), null);
+                else
+                    return Pair.create(Bits.ALL, null);
+
+            if (ordinalsToRowIds == null)
+                return Pair.create(Bits.ALL, null);
 
-        return new IgnoringBits(ignoredOrdinals, graph.size());
+            // Since we have a one to many mapping from ordinal to row id, we need to ensure we only ignore fully
+            // shadowed ordinals, and we also want to store the non-shadowed row ids to prevent having to read them
+            // from disk later.
+            try (var rowIdsView = graph.getRowIdsView())
+            {
+                for (var entry : ordinalsToRowIds.entrySet())
+                {
+                    int ordinal = entry.getKey();
+                    int[] shadowedRowIds = entry.getValue().toArray();
```

Comment:
.toArray() here copies the array (System.arraycopy).
Can this be avoided?

Reply:
Thanks for asking, I thought I needed this, but it was actually trivial to remove.